### PR TITLE
Fixed SAIC-70: VM verification after migration

### PR DESCRIPTION
**Pull request by realsystem**

### CODE REVIEW

File: cloudferrylib/os/actions/verify_vms.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        search_opts.update(kwargs.get('search_opts_tenant', {}))
+        compute_resource = self.cloud.resources[utl.COMPUTE_RESOURCE]
+        src_info = compute_resource.read_info(**search_opts)
+        src_inst = False
+        for src_inst_id in src_info['instances']:
+            for dst_inst in dst_info['instances'].values():
+                old_id = dst_inst['meta']['old_id']
+                if old_id == src_inst_id:
+                    src_inst = True
+                    continue
```
Comment (Settis) [inline cloudferrylib/os/actions/verify_vms.py]: I think you must use `break` here.

File: cloudferrylib/os/actions/verify_vms.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        if not kwargs.get('info'):
+            raise Exception("No info")
+        dst_info = kwargs['info']
+        search_opts = {'search_opts': kwargs.get('search_opts', {})}
+        search_opts.update(kwargs.get('search_opts_tenant', {}))
+        compute_resource = self.cloud.resources[utl.COMPUTE_RESOURCE]
+        src_info = compute_resource.read_info(**search_opts)
+        src_inst = False
+        for src_inst_id in src_info['instances']:
+            for dst_inst in dst_info['instances'].values():
```
Comment (Settis) [inline cloudferrylib/os/actions/verify_vms.py]: You can make a set from 'old_id' and just check:
`if src_inst_id in old_id_set: `
It's improve speed from O(n^2) to O(n).

**Comment by Settis** [comment]
+1, but you still need to rebase it.

File: cloudferrylib/os/actions/verify_vms.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an AS IS BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and#
+# limitations under the License.
+
+
+from cloudferrylib.base.action import action
+from cloudferrylib.utils import utils as utl
```
Comment (miarmak) [inline cloudferrylib/os/actions/verify_vms.py]: Why do we need to use `utl`? Let us stay with `utils`...

File: cloudferrylib/os/actions/verify_vms.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+
+class VerifyVms(action.Action):
+    def __init__(self, init, cloud=None):
```
Comment (miarmak) [inline cloudferrylib/os/actions/verify_vms.py]: I guess we can get rid of this `__init__` method, cause there are no any actions here (and just use `__init__` of parent class)

File: cloudferrylib/os/actions/verify_vms.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        super(VerifyVms, self).__init__(init, cloud)
+
+    def run(self, **kwargs):
+        if not kwargs.get('info'):
+            raise Exception("No info")
```
Comment (miarmak) [inline cloudferrylib/os/actions/verify_vms.py]: Let's use more specific exception, f.e. those one, that we have in `cloudferrylib/base/exception.py` (`AbortMigrationError`) and more detailed message.

File: cloudferrylib/os/actions/verify_vms.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        dst_info = kwargs['info']
+        search_opts = {'search_opts': kwargs.get('search_opts', {})}
+        search_opts.update(kwargs.get('search_opts_tenant', {}))
+        compute_resource = self.cloud.resources[utl.COMPUTE_RESOURCE]
+        src_info = compute_resource.read_info(**search_opts)
+        old_ids = set(dst_inst['meta']['old_id']
+                      for dst_inst in dst_info['instances'].values())
+        for src_inst_id in src_info['instances']:
+            if src_inst_id not in old_ids:
+                raise Exception("Instance was not migrated: {}".format(src_inst_id))
```
Comment (miarmak) [inline cloudferrylib/os/actions/verify_vms.py]: The same issue with exception, as above.

File: devlab/tests/scenarios/cold_migrate.yaml [hunk 7]
```diff
@@ -96,3 +96,5 @@ process:
       - save_result_migrate_instances: True
       - is_instances: ['get_next_instance']
       - rename_info_iter: True
+  - verify:
+      - verify_vms: True
```
Comment (miarmak) [inline devlab/tests/scenarios/cold_migrate.yaml]: What about `scenario/migrate.yaml`?

File: cloudferrylib/os/actions/verify_vms.py (hunk 6 above)
Comment (miarmak) [inline cloudferrylib/os/actions/verify_vms.py]: And also, there is a question: in this case we will fail, rollback will start and we'll lose all our migrated items. Is it an expected behaviour?

Comment (realsystem) [inline cloudferrylib/os/actions/verify_vms.py]: Yes. Warning in this case will be ignored by user.

File: cloudferrylib/os/actions/verify_vms.py (hunk 5 above)
Comment (mirrorcoder) [inline cloudferrylib/os/actions/verify_vms.py]: +1

File: cloudferrylib/os/actions/verify_vms.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        dst_info = kwargs['info']
+        search_opts = {'search_opts': kwargs.get('search_opts', {})}
+        search_opts.update(kwargs.get('search_opts_tenant', {}))
+        compute_resource = self.cloud.resources[utils.COMPUTE_RESOURCE]
+        src_info = compute_resource.read_info(**search_opts)
+        old_ids = set(dst_inst['meta']['old_id']
+                      for dst_inst in dst_info['instances'].values())
+        for src_inst_id in src_info['instances']:
+            if src_inst_id not in old_ids:
+                raise exception.AbortMigrationError(
```
Comment (roman-verchikov) [inline cloudferrylib/os/actions/verify_vms.py]: I think would be better to list VMs which failed to migrate, not the first one. Maybe incorrectly migrated VMs can be restored/updated manually if there are just few.
Also, simply comparing instance IDs is obviously not enough for such a verification.

**Comment by miarmak** [comment]
Please rebase.

And also, regarding Roman's coment above. Maybe really it is make sense not to make rollback of ALL migrated VMs, if  there are just few have been failed? Probably by providing any allowable percentage of failing without rollback successfully migrated VMs (and then try to restore/update/migrate failed VMs manually or via another one migration iteration)

File: scenario/tasks.yaml [hunk 9]
```diff
@@ -104,3 +104,4 @@ tasks:
    check_qemu_version: ['CheckQemuVersion']
    sleep_30: ['Sleep30']
    check_networks: ['CheckNetworks']
+   verify_vms: ['VerifyVms', src_cloud]
```
Comment (roman-verchikov) [inline scenario/tasks.yaml]: I'd expect we verify destination cloud, not source

File: cloudferrylib/os/actions/verify_vms.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+
+from cloudferrylib import utils
+from cloudferrylib.base import exception
+from cloudferrylib.base.action import action
+
+LOG = utils.get_log(__name__)
+
+
+class VerifyVms(action.Action):
```
Comment (roman-verchikov) [inline cloudferrylib/os/actions/verify_vms.py]: Please add docstring

File: scenario/tasks.yaml (hunk 9 above)
Comment (realsystem) [inline scenario/tasks.yaml]: We are verifying SRC and DST clouds

**Comment by roman-verchikov** [comment]
+1, let's wait for functional tests

**Comment by roman-verchikov** [comment]
@realsystem please extend with following verifications:

* All VMs migrated;
* VMs have following:
    * Connected to the same network(s);
    * Have the same keypairs associated;
    * Have the same image;
    * Have the same flavor;
    * Have the same floating IP associated (if any)

File: cloudferrylib/os/actions/verify_vms.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+                src_inst_info = src_info['instances'][src_inst_id]['instance']
+                src_cmp_info[src_inst_id].append(src_inst_info['name'])
+                src_cmp_info[src_inst_id].append(src_inst_info['flav_details'])
+                src_cmp_info[src_inst_id].append(src_inst_info['key_name'])
+                src_cmp_info[src_inst_id].append(src_inst_info['interfaces'])
+                src_cmp_info[src_inst_id].append(src_inst_info['volumes'])
+        if failed_vms or (dst_cmp_info != src_cmp_info):
+            LOG.warning("Instances were not migrated:")
+            for vm in failed_vms:
+                LOG.warning("{}".format(vm))
```
Comment (roman-verchikov) [inline cloudferrylib/os/actions/verify_vms.py]: Please increase debug logging so that end-user can understand what's being verified.

**Comment by miarmak** [comment]
Unit tests have failed with the next error:
```
1) ERROR: tests.cloudferrylib.os.actions.test_verify_vms.VerifyVmsTest.test_verify_vms_no_exc

   Traceback (most recent call last):
   _StringException: Traceback (most recent call last):
     File "/home/jenkins/workspace/cloudferry-unittests-and-style-checks/CloudFerry/tests/cloudferrylib/os/actions/test_verify_vms.py", line 67, in test_verify_vms_no_exc
       action.run(info=self.fake_dst_info)
     File "/home/jenkins/workspace/cloudferry-unittests-and-style-checks/CloudFerry/cloudferrylib/os/actions/verify_vms.py", line 42, in run
       dst_inst_ = dst_inst['instance']
   KeyError: 'instance'
```
Fix it please.

File: cloudferrylib/os/actions/verify_vms.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+                new_dst_volumes.append(dst_vol['volume'])
+            dst_cmp_info[old_id].append(new_dst_volumes)
+            inst_cnt += 1
+        failed_vms = []
+        src_cmp_info = {}
+        for src_inst_id in src_info['instances']:
+            if src_inst_id not in old_ids:
+                failed_vms.append(src_inst_id)
+            else:
+                src_cmp_info[src_inst_id] = []
```
Comment (miarmak) [inline cloudferrylib/os/actions/verify_vms.py]: In case of using `List` structure we are dependent on it's order...maybe it is better to use `Dict` here?

File: cloudferrylib/os/actions/verify_vms.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+                src_inst_info = src_info['instances'][src_inst_id]['instance']
```
Comment (miarmak) [inline cloudferrylib/os/actions/verify_vms.py]: And also, filling these structures (for src and dst) seems pretty similar. Maybe it is possible to refactor it and place in the another method?

Comment (realsystem) [inline cloudferrylib/os/actions/verify_vms.py]: What is the purpose of this change?

File: cloudferrylib/os/actions/verify_vms.py (hunk 12 above)
Comment (realsystem) [inline cloudferrylib/os/actions/verify_vms.py]: Which case can break the current logic?

Comment (miarmak) [inline cloudferrylib/os/actions/verify_vms.py]: For example adding any new attributes to check. We'll need to follow the order in 2 different parts of code. 
But it is just a question and if you feel that it's not an issue, so we can stay with the current code.

File: cloudferrylib/os/actions/verify_vms.py (hunk 13 above)
Comment (miarmak) [inline cloudferrylib/os/actions/verify_vms.py]: F.e. we need to change smth, or add or delete. We'll need to do the same stuff in 2 different code blocks. (simple duplicate code).
But the same as above about question and the advisability of this change.

Comment (realsystem) [inline cloudferrylib/os/actions/verify_vms.py]: Got it.

**Comment by roman-verchikov** [comment]
Looks good to me, @ntymtsiv please verify

**Comment by miarmak** [comment]
Looks good for me too. +1

File: cloudferrylib/os/actions/verify_vms.py [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+                    {'name': src_inst_info['name']})
+                src_cmp_info[src_inst_id].update(
+                    {'flav_details': src_inst_info['flav_details']})
+                src_cmp_info[src_inst_id].update(
+                    {'key_name': src_inst_info['key_name']})
+                src_cmp_info[src_inst_id].update(
+                    {'interfaces': src_inst_info['interfaces']})
+                src_cmp_info[src_inst_id].update(
+                    {'volumes': src_inst_info['volumes']})
+        if failed_vms or (dst_cmp_info != src_cmp_info):
```
Comment (ntymtsiv) [inline cloudferrylib/os/actions/verify_vms.py]: In case when second condition is True (dst_cmp_info != src_cmp_info) we have warning message "Instances were not migrated:" and that's all. I think this check should display more useful information in this case.

**Comment by ntymtsiv** [comment]
In case when we have assigned floating ip to vm, we have next dicts:
src['a3b30399-8b96-464f-b09c-fdbd19c70630']
{'flav_details': {'ephemeral_gb': 0,
  'memory_mb': 48,
  'name': u'flavorname2',
  'root_gb': 2,
  'vcpus': 2},
 'interfaces': [{'floatingip': None,
   'ip': u'22.2.2.4',
   'mac': u'fa:16:3e:75:93:d3',
   'name': u'tenantnet2'}],
 'key_name': u'key2',
 'name': u'keypair_test_server',
 'volumes': []}
dst['a3b30399-8b96-464f-b09c-fdbd19c70630']
{'flav_details': {'ephemeral_gb': 0,
  'memory_mb': 48,
  'name': u'flavorname2',
  'root_gb': 2,
  'vcpus': 2},
 'interfaces': [{'floatingip': u'192.168.1.104',
   'ip': u'22.2.2.4',
   'mac': u'fa:16:3e:75:93:d3',
   'name': u'tenantnet2'}],
 'key_name': u'key2',
 'name': u'keypair_test_server',
 'volumes': []}

File: cloudferrylib/os/actions/verify_vms.py [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+                src_cmp_info[src_inst_id] = {}
+                src_inst_info = src_info['instances'][src_inst_id]['instance']
+                src_cmp_info[src_inst_id].update(
+                    {'name': src_inst_info['name']})
+                src_cmp_info[src_inst_id].update(
+                    {'flav_details': src_inst_info['flav_details']})
+                src_cmp_info[src_inst_id].update(
+                    {'key_name': src_inst_info['key_name']})
+                src_cmp_info[src_inst_id].update(
+                    {'interfaces': src_inst_info['interfaces']})
```
Comment (ntymtsiv) [inline cloudferrylib/os/actions/verify_vms.py]: floatingip already removed from src info

**Comment by roman-verchikov** [comment]
can be merged.
